src/io/delimited.cr

```diff
@@ -46,6 +46,147 @@ class IO::Delimited < IO
+        if peek.size >= delimiter_remaining.size
+          # We found the delimiter!
+          @io.skip(min_size)
+          @finished = true
+          return 0
+        else
+          # Copy the remaining of peek to the active delimiter buffer for now
+          (@delimiter_buffer + @active_delimiter_buffer.size).copy_from(peek)
+          @active_delimiter_buffer = @delimiter_buffer[0, @active_delimiter_buffer.size + peek.size]
+
+          # Skip whatever we had in peek, and try reading more
+          @io.skip(peek.size)
+          return read_internal(slice)
+        end
+      else
+        # No match: just copy from the active delimiter
+        read_bytes = @active_delimiter_buffer.size
+        slice.copy_from(@active_delimiter_buffer)
+        slice += read_bytes
+        @active_delimiter_buffer = Bytes.empty
+        return read_bytes + read_internal(slice)
+      end
+    end
+
+    first_byte = @read_delimiter[0]
+    index = peek.index(first_byte)
+
+    # If we can't find the delimiter's first byte we can just read from peek
+    unless index
+      # If we have more in peek than what we need to read, read all of that
+      if peek.size >= slice.size
+        slice.copy_from(peek[0, slice.size])
+        @io.skip(slice.size)
+        return slice.size
+      else
+        # Otherwise, read from peek for now
+        slice.copy_from(peek)
+        @io.skip(peek.size)
+        slice += peek.size
+        return peek.size
+      end
+    end
+
+    # If the matching byte is past the data we need to read,
+    # we can fill the entire slice
+    if index >= slice.size
+      slice.copy_from(peek[0, slice.size])
+      @io.skip(slice.size)
+      return slice.size
+    end
+
+    # If the delimiter is just a single byte, we can stop right here
+    if @delimiter_buffer.size == 1
+      slice.copy_from(peek[0, index])
+      @io.skip(index + 1)
+      @finished = true
+      return index
+    end
+
+    # If the delimiter fits the rest of the peek buffer,
+    # we can check it right now.
+    if index + @delimiter_buffer.size <= peek.size
+      # If we found the delimiter, we are done
+      if peek[index, @delimiter_buffer.size] == @read_delimiter
+        slice.copy_from(peek[0, index])
+        @io.skip(index + @delimiter_buffer.size)
+        @finished = true
+        return index
+      else
+        # Otherwise, we can read up to past that byte for now
+        slice.copy_from(peek[0, index + 1])
+        @io.skip(index + 1)
+        slice += index + 1
+        return index + 1
+      end
+    end
+
+    # If the part past in the peek buffer past the matching index
+    # doesn't match the read delimiter's portion, we can move on
+    rest = peek[index..]
+    unless rest == @read_delimiter[0, rest.size]
+      # If not, we can read from the peek buffer for now
+      slice.copy_from(peek)
+      @io.skip(peek.size)
+      slice += peek.size
+      return peek.size
+    end
```

Comment:
At this point, we only know about `rest` that it can't hold the entire of the delimiter. But there is no reason to assume it can't contain the start of the delimiter at a different position after `index`.

```
 delimiter
    --
   aab
   ---
  rest
```

This is similar to the previously mentioned edge case with part of a duplicate sequence delimiter in the active buffer.

Reply:
@straight-shoota Is this still a remaining concern? Do I understand correctly that you think there might be a bug?

Reply:
Oooh, yeah I think so. Dismissing my approval.

I think this needs the same handling as the previous case, with only consuming up to the `index`, not the entire peek buffer.